extensions/auth/portability-auth-microsoft/src/main/java/org/dataportabilityproject/auth/microsoft/MicrosoftAuthServiceExtension.java

```diff
@@ -1,41 +1,74 @@
 package org.dataportabilityproject.auth.microsoft;
 
 import com.google.common.collect.ImmutableList;
+import java.io.IOException;
 import java.util.List;
+import okhttp3.OkHttpClient;
 import org.dataportabilityproject.api.launcher.ExtensionContext;
+import org.dataportabilityproject.spi.cloud.storage.AppCredentialStore;
 import org.dataportabilityproject.spi.gateway.auth.AuthDataGenerator;
-import org.dataportabilityproject.spi.gateway.auth.extension.AuthServiceExtension;
 import org.dataportabilityproject.spi.gateway.auth.AuthServiceProviderRegistry.AuthMode;
+import org.dataportabilityproject.spi.gateway.auth.extension.AuthServiceExtension;
+import org.dataportabilityproject.types.transfer.auth.AppCredentials;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /** */
 public class MicrosoftAuthServiceExtension implements AuthServiceExtension {
-  private MicrosoftAuthDataGenerator contactsGenerator;
+  private static final ImmutableList<String> DEFAULT_AUTH_GENERATOR_SERVICES =
+      ImmutableList.of("contacts", "calendar");
+  private static final Logger logger = LoggerFactory.getLogger(MicrosoftAuthServiceExtension.class);
 
-  public MicrosoftAuthServiceExtension() {
-    // contactsGenerator = new MicrosoftAuthDataGenerator(r)
-  }
+  private volatile MicrosoftAuthDataGenerator authDataGenerator;
+
+  public MicrosoftAuthServiceExtension() {}
 
   public String getServiceId() {
     return "microsoft";
   }
 
   public AuthDataGenerator getAuthDataGenerator(String transferDataType, AuthMode mode) {
-    // TODO
-    throw new UnsupportedOperationException();
+    // TODO Create auth data generator for a given mode, usually means a different scope
+    if (DEFAULT_AUTH_GENERATOR_SERVICES.contains(transferDataType)) {
+      return authDataGenerator;
+    }
+    throw new UnsupportedOperationException("transferDataType does not have an AuthDataGenerator");
   }
 
   @Override
   public List<String> getImportTypes() {
-    return ImmutableList.of();
+    return DEFAULT_AUTH_GENERATOR_SERVICES;
   }
 
   @Override
   public List<String> getExportTypes() {
-    return ImmutableList.of();
+    return DEFAULT_AUTH_GENERATOR_SERVICES;
   }
 
   @Override
   public void initialize(ExtensionContext context) {
-    // TODO
+    makeAuthDataGenerator(context);
+  }
+
+  private synchronized MicrosoftAuthDataGenerator makeAuthDataGenerator(ExtensionContext context) {
+    if (authDataGenerator == null) {
+      AppCredentials credentials;
+      try {
+        credentials = context.getService(AppCredentialStore.class)
```

Comment:
discussed offline - I think of AppCredentialStore being a property of the CloudExtension and that certain extensions (eg AuthServiceExtension and TransferExtension) should have a notion that they are running on "a cloud" and thus have the CloudExtension available to them. Perhaps through an extra param to Extension.initialize(context, CloudExtension). Or we could make the concept of extensions using other extensions generic with Extension.initialize(context, Extension ... extensions). Not sure if this has come up before or if others have thoughts

Reply:
Another option would be for ExtensionContext to have a getExtension method and have ExtensionContext responsible for service loading

Reply:
Discussed offline - It makes sense to have different extensions unaware of the other extensions provided. So given this, it is best to have different ExtensionContext passed into the different extensions.